x-pack/legacy/plugins/apm/server/routes/typings.ts

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import t, { Type, HasProps, TypeOf } from 'io-ts';
+import { Request } from 'hapi';
+import { InternalCoreSetup } from 'src/core/server';
+
+type ExtractTType<T> = T extends Type<any, any, any> ? TypeOf<T> : {};
+
+export type ExtractTTypeFromParams<T extends RouteParams> = {
+  [key in keyof T]: ExtractTType<T[key]>;
+};
+
+interface RouteType<
+  T extends Partial<Record<keyof RouteParams, Type<any, any, any>>>,
+  U extends any
+> {
+  params: T;
+  returnType: U;
+}
+
+export interface RouteDescriptionTree {
+  [key: string]: {
+    [method in HttpMethod]: RouteType<any, any>;
+  };
+}
+
+export interface RouteParams {
+  path?: HasProps;
+  query?: HasProps;
+  body?: HasProps | Type<any>;
+}
+
+export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'DELETE';
+
+export type RouteRequestHandler<T extends RouteParams, U> = (
+  req: Request,
+  params: ExtractTTypeFromParams<T>
+) => Promise<U>;
+
+export interface RouteDescription<
+  T extends string,
+  U extends HttpMethod,
+  V extends RouteParams,
+  W
+> {
+  path: T;
+  method: U;
+  params?: V;
+  handler: RouteRequestHandler<V, W>;
```

Comment:
Not blaming you: It's a bit of a mystery why the convention in typescript is to use short and cryptic names for generic parameters.
Maybe I'm just the odd one out but what about:
```ts
path: Path;
method: Method;
params?: Params;
```

Reply:
(same goes for `W` which I guess is `Response` but genuinely can't figure it out at a glance)

Reply:
Changed this to `T{Type}` per our discussion.